rs/dre-canisters/trustworthy-node-metrics/src/trustworthy-node-metrics/src/metrics_manager.rs

```diff
@@ -167,24 +140,121 @@ fn grouped_by_node(subnet_metrics: Vec<(PrincipalId, Vec<NodeMetricsHistoryRespo
     grouped_by_node
 }
 
-async fn update_node_providers(nodes_principal: Vec<&PrincipalId>) {
+fn generate_node_type(node_types_count: Option<BTreeMap<String, i32>>, mut rewardable_nodes: BTreeMap<String, u32>) -> String {
+    if let Some(node_types) = &node_types_count {
+        for (node_type, decrement_value) in node_types.iter() {
+            if let Some(reward_count) = rewardable_nodes.get_mut(node_type) {
+                *reward_count = reward_count.saturating_sub(*decrement_value as u32);
+            }
+        }
+    }
+
+    if rewardable_nodes.is_empty() {
+        "unknown:no_rewardable_nodes_found".to_string()
+    } else {
+        loop {
+            let (k, v) = match rewardable_nodes.pop_first() {
+                Some(kv) => kv,
+                None => break "unknown:rewardable_nodes_used_up".to_string(),
+            };
+            if v != 0 {
+                break k;
```

Comment:
Don't suggest using `break` because here it is meant to actually `return`. Its misleading. `Break exp` is just a way to be able to easily assign a value from a loop, so you don't have to use `mut temp...`